### PR TITLE
Update Sparse_Segtree.mdx

### PR DESCRIPTION
**Pull request by eysbutno**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

AC:
https://oj.uz/submission/1104138
https://oj.uz/submission/1104136

- added (IMO cleaner) implementations for a pointer and (vector based) indice sparse segtree
- added time complexity and cleaned up the module a little bit

lmk if I trolled a bit on the indice impl or there are some other issues

### CODE REVIEW

File: content/5_Plat/Sparse_Segtree.mdx [hunk 1]
```diff
@@ -63,130 +64,196 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
+		Node *left = nullptr;
+		Node *right = nullptr;
+	};
+	Node *root = new Node;
+	const int n;
+
+	int comb(int a, int b) { return a + b; }
+
+	void apply(Node *cur, int len, int val) {
+		if (val == 1) {
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:92]: screaming, begging for you to rename `cur` to `curr`

File: content/5_Plat/Sparse_Segtree.mdx [hunk 2]
```diff
@@ -63,130 +64,196 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
+</CPPSection>
+</LanguageSection>
 
-		if (l > mid) update(segtree[node].r, l, r);
-		else if (r <= mid) update(segtree[node].l, l, r);
-		else {
-			update(segtree[node].l, l, mid);
-			update(segtree[node].r, mid + 1, r);
-		}
+### Indice Implementation
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:157]:
```suggestion
### Index Implementation
```

File: content/5_Plat/Sparse_Segtree.mdx [hunk 3]
```diff
@@ -63,130 +64,196 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
 
-	segtree[1].sum = 0;
-	segtree[1].lazy = 0;
-	segtree[1].tl = 1;
-	segtree[1].tr = 1e9;
+	int range_sum(int ql, int qr) { return range_sum(0, 0, n - 1, ql, qr); }
+};
 
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:236]: is this necessary

File: content/5_Plat/Sparse_Segtree.mdx [hunk 4]
```diff
@@ -63,130 +64,196 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
 ```
 
-</CPPSection>
-
-</LanguageSection>
-
 <Optional>
 
-It's possible to reduce the memory of a sparse segment tree to $\mathcal{O}(N)$,
+It's possible to reduce the memory of a sparse segment tree to $\mathcal{O}(Q)$,
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:256]: lmao

File: content/5_Plat/Sparse_Segtree.mdx [hunk 5]
```diff
@@ -63,130 +64,196 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
+		if (qr < l || ql > r) { return 0; }
+		if (ql <= l && r <= qr) { return cur->freq; }
+		push_down(cur, l, r);
+		int m = (l + r) / 2;
+		return comb(range_sum(cur->left, l, m, ql, qr),
+		            range_sum(cur->right, m + 1, r, ql, qr));
+	}
+
+  public:
+	SparseSegtree(int n) : n(n) {}
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:129]: what's n here

File: content/5_Plat/Sparse_Segtree.mdx [hunk 6]
```diff
@@ -63,130 +64,196 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
 	}
-}
 
-int main() {
-	iostream::sync_with_stdio(false);
-	cin.tie(0);
-	int m;
-	cin >> m;
+  public:
+	SparseSegtree(int n, int q) : n(n) {
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:225]: what's n & q here

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Sparse_Segtree.mdx (hunk 3 above)
Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx:236]: shoot forgot to take it out, mb

File: content/5_Plat/Sparse_Segtree.mdx (hunk 6 above)
Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx:225]: n = range size, q = # of queries

the module explanation uses those variables I'm pretty sure

File: content/5_Plat/Sparse_Segtree.mdx (hunk 4 above)
Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx:255]: $N$ is used to refer to the range size here so the prev version was wrong

File: content/5_Plat/Sparse_Segtree.mdx (hunk 1 above)
Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx:92]: wait why

i dont know anyone else who does that LOL

File: content/5_Plat/Sparse_Segtree.mdx [hunk 7]
```diff
@@ -63,130 +64,195 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
-	segtree[1].lazy = 0;
-	segtree[1].tl = 1;
-	segtree[1].tr = 1e9;
+	int range_sum(int ql, int qr) { return range_sum(0, 0, n - 1, ql, qr); }
+};
 
+int main() {
+	int n;
+	cin >> n;
+	SparseSegtree st((int)(1e9 + 1), n);
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:238]: magic number?

Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx:238]: wdym

1e9 + 1 is for the range $[1, 10^9]$ that is specified in the problem

Comment (SansPapyrus683) [reply content/5_Plat/Sparse_Segtree.mdx]: put it in a constant

File: content/5_Plat/Sparse_Segtree.mdx [hunk 8]
```diff
@@ -235,7 +236,8 @@ class SparseSegtree {
 int main() {
 	int query_num;
 	cin >> query_num;
-	SparseSegtree st((int)(1e9 + 1), query_num);
+	const int range_size = 1e9 + 1;
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:239]: is the range inclusive or exclusive

Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx:239]: like all the other ds impls in the guide it's $[0, n)$ where $n$ is the size given

File: content/5_Plat/Sparse_Segtree.mdx [hunk 9]
```diff
@@ -63,130 +62,199 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
+
+	void range_set(int ql, int qr, int val) { range_set(root, 0, n - 1, ql, qr, val); }
+
+	int range_sum(int ql, int qr) { return range_sum(root, 0, n - 1, ql, qr); }
+};
+
+int main() {
+	int query_num;
+	cin >> query_num;
+	const int range_size = 1e9;
```
Comment (ryanchou-dev) [inline content/5_Plat/Sparse_Segtree.mdx:137]: capitalize?

**Review by ryanchou-dev — COMMENTED** [review]
cleanest code 😮

File: content/5_Plat/Sparse_Segtree.mdx (hunk 9 above)
Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx]: good point

**Comment by TheGamingMousse** [comment]
> cleanest code 😮

🤥

File: content/5_Plat/Sparse_Segtree.mdx [hunk 10]
```diff
@@ -63,130 +62,212 @@ range is up to $10^9$, so we have to use a sparse segment tree.
...
+<Warning>
 
-		if (l > mid) update(segtree[node].r, l, r);
-		else if (r <= mid) update(segtree[node].l, l, r);
-		else {
-			update(segtree[node].l, l, mid);
-			update(segtree[node].r, mid + 1, r);
-		}
+Instantiating multiple pointer segment trees could result in a memory leak
+because C++ doesn't have a garbage collector.
```
Comment (SansPapyrus683) [inline content/5_Plat/Sparse_Segtree.mdx:157]: i mean even instantiating one is kinda a memory leak since you don't do any cleaning up

Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx:157]: yeah but it's only problematic if ur running thru multiple test cases and use multiple segtrees

idk, it's typically not an issue but something good to note imo

Comment (SansPapyrus683) [reply content/5_Plat/Sparse_Segtree.mdx:157]: hm hb "note that this impl isn't gc'd, so if you redclare stuff it's kinda cooked"

Comment (TheGamingMousse) [reply content/5_Plat/Sparse_Segtree.mdx]: k

**Review by SansPapyrus683 — APPROVED** [review]
hello everybody my name is markiplier and welcome back to

**Review by envyaims — APPROVED** [review]
nice!! this is actually rly good

**Pull request merged**